src/metabase/api/table.clj

```diff
@@ -551,4 +552,24 @@
                 :file     (get-in raw-params ["file" :tempfile])
                 :action   ::upload/replace}))
 
+(defn- delete-csv!
+  "This helper function exists to make testing the DELETE /api/table/:id endpoint easier."
+  [table-id]
+  (try
+    (let [table  (api/check-404 (t2/select-one :model/Table :id table-id))
+          result (upload/delete-upload! table)]
+      {:status 200
+       :body   (= :done result)})
+    (catch Throwable e
+      {:status (or (-> e ex-data :status-code)
+                   500)
+       :body   {:message (or (ex-message e)
+                             (tru "There was an error deleting the table"))}})))
+
+(api/defendpoint ^:multipart DELETE "/:id"
```

Comment:
I also wonder if we want to make this API more narrowed, something like `DELETE /api/delete/:id/delete-csv`. The pros are:
- it follows the same pattern as all of the csv APIs
- preserve `DELETE /api/delete/:id` for future usages of deleting normal tables.

Reply:
I thought about this, but decided this route made more sense:

- The other routes only support uploading CSV files, but we may support other formats in future.
- This route would apply to all uploads.
- The fact that a table is an upload is a property of the table, i.e. implicit in the id.
- If we want to add support for deleting other tables, which seems unlikely, I don't see an issue reusing this route.
- In REST a DELETE points at the corresponding resource, I don't see a reason for an idiosyncratic RPC URL.

With respect to the last point, it could have been nice to have these routes under `/uploads/tables/:id` to have implicit filtering, but I don't think it's worth migrating or supporting both at this time.

Reply:
Since we ended up moving the feature to an EE module, it ended up being scoped to uploads after all.